tracing-core/src/metadata.rs

```diff
@@ -180,6 +180,35 @@ impl<'a> Metadata<'a> {
         }
     }
 
+    /// Construct new metadata for a span or event, with a name, target, level, field
+    /// names, and optional source code location using dynamically allocated data.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(any(feature = "std", feature = "alloc"))))]
+    pub fn from_cow<S>(
+        name: S,
+        target: S,
+        level: Level,
+        file: Option<S>,
+        line: Option<u32>,
+        module_path: Option<S>,
+        fields: field::FieldSet,
+        kind: Kind,
+    ) -> Self
+    where
+        S: Into<Cow<'a, str>>,
```

Comment:
Using a single type parameter `S: Into<Cow<'a, str>>` requires that _all_ of these values be of the same type. This means that you couldn't pass a mix of `&'static str`s and `String`s (for example). 

Can we use `impl Into<Cow<'a, str>>` instead? This should allow passing parameters of different types, as long as they all implement `Into<Cow<'a, str>>`
```suggestion
    pub fn from_cow(
        name: impl Into<Cow<'a, str>>,
        target: impl Into<Cow<'a, str>>,
        level: Level,
        file: Option<impl Into<Cow<'a, str>>>,
        line: Option<u32>,
        module_path: Option<impl Into<Cow<'a, str>>>,
        fields: field::FieldSet,
        kind: Kind,
    ) -> Self
```

Reply:
You're right, I think we have to use `impl Into<…`. 

The downside is that we need type annotations sometimes for `file` and `module_path` when calling `from_cow("a name", "a target", lvl, None::<Cow<'_, str>, None, "a mod path", fields, Kind::EVENT)`.

Reply:
To elaborate on this, last night I was toying with the idea of having explicit constructors, i.e. `Metadata::from_string` and `Metadata::from_str`. It would help with the naming problem, but other than that it's not a very fruitful idea.

Reply:
Explicit constructors could be fine, but again, the issue is that it would require the provided values to either all be `&str` or all be `String`. I think `impl Trait` is the only approach that allows an arbitrary mix of string slices and strings.

I think the type inference issue _might_ be fixed if we changed the `Option`al params from `Option<impl Into<Cow<'a, str>>>` to `impl Into<Option<Cow<'a, str>>>`, but i'm not sure if that would actually change anything. `None`s might always have to be annotated?

Reply:
>  but i'm not sure if that would actually change anything. `None`s might always have to be annotated?

You are correct that `impl Option<Cow<…>>` lets us pass `None` without annotations, but I can't make it work nicely for the `Some` case: callers now have to convert to `Cow` at the callsite, `Metadata::from_cow(…, Some("a file".into()), …)`. Kind of defeats the purpose I think.

I get this:
```
187 |     pub fn from_cow(
    |            -------- required by a bound in this
...
191 |         file: impl Into<Option<Cow<'a, str>>>,
    |                    -------------------------- required by this bound in `metadata::Metadata::<'a>::from_cow`
...
954 |         let _metadata = Metadata::from_cow(
    |                         ^^^^^^^^^^^^^^^^^^ the trait `From<Option<&str>>` is not implemented for `Option<Cow<'_, str>>`
    |
    = help: the following implementations were found:
              <Option<&'a T> as From<&'a Option<T>>>
              <Option<&'a mut T> as From<&'a mut Option<T>>>
              <Option<T> as From<T>>
```

Between the two I prefer annotating the `None`s.

Reply:
Oh well, it's too bad that didn't work out. Let's either use `Option<into Cow<...>>`, or  just take `Option<Cow<...>>` and make the user do the conversion, I guess.

Reply:
I think keeping `Option<impl Into<Cow<'a, str>>>` is the least worst option. Yes, it becomes really annoying to pass a `None`, but the compiler might improve and if it does we have the right API in place. Do you have any intuition for how common it is to not have a file and a module path in production code?

Reply:
I think if a user is explicitly calling a `Metadata` constructor (rather than using a macro), we're already in a pretty uncommon case? The macros will always have source code locations (files, lines, and module paths). I think these will probably only be `None` when consuming traces from other systems that might not have Rust source locations associated with them?